builder/dockerfile/copy.go

```diff
@@ -288,33 +289,56 @@ func walkSource(source builder.Source, origPath string) ([]string, error) {
 	return subfiles, nil
 }
 
-type sourceDownloader func(string) (builder.Source, string, error)
+type sourceDownloader func(url, dest string) (builder.Source, string, error)
 
 func newRemoteSourceDownloader(output, stdout io.Writer) sourceDownloader {
-	return func(url string) (builder.Source, string, error) {
-		return downloadSource(output, stdout, url)
+	return func(url, dest string) (builder.Source, string, error) {
+		return downloadSource(output, stdout, url, dest)
 	}
 }
 
-func errOnSourceDownload(_ string) (builder.Source, string, error) {
+func errOnSourceDownload(_, _ string) (builder.Source, string, error) {
 	return nil, "", errors.New("source can't be a URL for COPY")
 }
 
-func downloadSource(output io.Writer, stdout io.Writer, srcURL string) (remote builder.Source, p string, err error) {
+func downloadSource(output io.Writer, stdout io.Writer, srcURL, dest string) (remote builder.Source, p string, err error) {
 	u, err := url.Parse(srcURL)
 	if err != nil {
 		return
 	}
-	filename := filepath.Base(filepath.FromSlash(u.Path)) // Ensure in platform semantics
-	if filename == "" {
-		err = errors.Errorf("cannot determine filename from url: %s", u)
-		return
+	var filename string
+	if u.Path != "" && !strings.HasSuffix(u.Path, "/") {
+		filename = filepath.Base(filepath.FromSlash(u.Path)) // Ensure in platform semantics
+		if filename == "" {
+			err = errors.Errorf("cannot determine filename from url: %s", u)
+			return
+		}
+	} else if dest != "" && !strings.HasSuffix(dest, "/") {
+		filename = filepath.Base(filepath.FromSlash(dest))
+		if filename == "" {
+			err = errors.Errorf("cannot determine filename from dest: %s", dest)
+			return
+		}
```

Comment:
Please split this into a new function:

```go
func getTargetFilename(url *url.Url) string {

}
```

I don't think we need to pass `dest` all the way from through function call chain. The filename doesn't seem terribly relevant. Can't we just use a random name, or even a hardcoded name since we're creating a new temporary directory anyway?

I think `getTargetFilename` should make a "best guess" at a filename (using `filepath.Base()` like it did before), and if it fails just returns the string `"download"` instead of erroring or using dest or content-dispoition.